---
title: "01_fig2F_h2Enrich_CTspecificGenes"
author: Shuyang Yao (shuyang.yao@ki.se)
output: html_document
date: "2024-09-12"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Purpose: perform heritability (h2) enrichment analysis for specifically expressed gene per cell type.  
Steps:  
1. filter genes based on expression.  
- filter: Keep transcripts with at least 1 count in at least 10% cells in each cell type.  
(i.e., remove transcripts with 0 expression in >90% cells in all cell types)
2. prepare annotations.  
- top 10% specific genes in each cell type.  
3. run h2 enrichment analysis.  
4. clean results.  
  
1. Filter genes  
```{r}
# R version: 4.1.0

library(tidyverse)
library(loomR)
library(data.table)

# input loom files
filepath <- "" # specify path to loom files; on loom file per cell type.
filelist <- grep(list.files(filepath, pattern = "loom"), pattern='downsampled', invert=TRUE, value=TRUE)

# working directory
WD <- "./7_heritability_enrichment/data" # specify working directory
setwd(WD)

#- logic of filtering:
# loop over cell type
# per cell type, get 
#- 1) per gene, count of cells with gene_count==0, count_cell_0exp, 
#- 2) total number of cells, n_total_cells.
# extract genes with count_cell_0exp > 0.9*n_total_cells
# output of loop, gene list: g2check, with: ENSGID, cluster, count_cell_0exp, N_cell_total; 
# This list is filtered for genes with zero expression in > 90% of cells.

for(i in 1:length(filelist)){
  file=paste(filepath, filelist[i], sep="")
  mycluster=basename(file)
  mycluster <- gsub("Samples_cellranger_pagoda_TH_and_D_adj_filtered_conos_cluster_based_","",mycluster)
  mycluster <- gsub(".loom","",mycluster)
  
  # link to loom file
  lfile <- connect(file,mode = "r", skip.validate = TRUE)
  # get expression_cell matrix
  exp <- as.data.frame(t(lfile[["matrix"]][,]))
  
  # number of zeros per row
  exp2 <- rowSums(exp==0)%>% as.data.frame()
  colnames(exp2) <- "count_cell_0exp"
  exp2$N_cell_total <- ncol(exp)
  rm(exp)
  exp2$ENSGID <- lfile[["row_attrs/Accession"]][]
  exp2$cluster <- mycluster
  
  # filter for genes with 0-expression in > 90% cells 
  mydf <- exp2 %>% filter(count_cell_0exp > 0.9*N_cell_total) %>%
    distinct(ENSGID, cluster)
  rm(exp2)
  # combine
  if(i==1){
    g2check <- mydf
    rm(mydf)
    print(i)
  }else{
    g2check <- rbind(g2check, mydf)
    rm(mydf)
    print(i)
  }
}

# take intersection of genes across all cell types
nr_cluster <- length(unique(g2check$cluster)) # get total number of clusters

g2rm <- g2check %>%
  distinct(ENSGID, cluster) %>%
  group_by(ENSGID) %>%
  summarise(n_cluster=n()) %>%
  filter(n_cluster==nr_cluster) # require the gene to remove having low expression in all cell types

# output
outpath <- paste(WD,"/g2rm_high0exp_all_CTs.tsv", sep="")
fwrite(g2rm, outpath, col.names = T, sep="\t")

#--- end ---#

```
  
2. Prepare annotation of specifically expressed genes.  
2.1 extract expression from loom files.  
```{r}
#= 2.1 extract expression from loom files
#= transcript count aggregated across cells per transcript (gene) per cluster

# input loom files
filepath <- "" # path to loom files
filelist <- grep(list.files(filepath, pattern = "loom"), pattern='downsampled', invert=TRUE, value=TRUE) 
# working directory
WD <- ""
setwd(WD)

# outputs:
# 1. duplicated genes
# 2. aggregated matrix: mean count 

for(i in 1:length(filelist)){
  file=paste(filepath, filelist[i], sep="")
  mycluster=basename(file)
  mycluster <- gsub("Samples_cellranger_pagoda_TH_and_D_adj_filtered_conos_cluster_based_","",mycluster)
  mycluster <- gsub(".loom","",mycluster)
  
  # link to loom file
  lfile <- connect(file,mode = "r", skip.validate = TRUE)
  # get expression_cell matrix
  exp <- as.data.frame(t(lfile[["matrix"]][,]))
  exp$rowsum <- rowSums(exp)
  exp$ENSGID <- lfile[["row_attrs/Accession"]][]
  
  # duplicate genes output: none
  
  # output Count
  exp.count.agg <- exp %>% select(ENSGID,rowsum) %>% 
    mutate(Expr=rowsum/(ncol(exp)-2), # ncol(exp)-2 is the number of cells
           cluster=mycluster) %>%
    select(-rowsum)
  countpath <- paste(WD,"/meanCount_", mycluster,".tsv", sep="")
  fwrite(exp.count.agg, countpath, col.names = T, sep="\t")
  
}

```
  
2.2 get top specific genes per cell type as the annotations.  
```{r}
#= 2.2 get top specific genes per cell type as the annotations.
#= Steps:
# 1. combine different cluster (rbind)
# 2. filter genes
#- remove duplicated genes (already removed in previous step)
#- remove genes with zero expression in >90% cells in all clusters
#- Keep only protein-coding genes from geneMatrix
# 3. normalize to 1 TPM per cluster
# 4. specificity
# 5. output top 10% as the annotation for h2 enrichment analysis

library(data.table)
library(tidyverse)

#- specify paths:
path_geneMx <- "" # path to geneMatrix
a <- "./7_heritability_enrichment/data" # path for aggregated expression per cluster (files meanCount_*.tsv from step 2.1)

#= 1. read in data from all cell types
b <- list.files(a, pattern = "meanCount_") %>% as.data.frame() 
colnames(b) <- "filename"
b <- b %>% mutate(full=paste(a,filename,sep="/"))
dat <- as.list(b$full) %>%
  map_df(function(x) fread(x, stringsAsFactors=FALSE, data.table=FALSE))

#= 2. filter genes
#- remove genes with zero expression in >90% cells in all clusters (from part 1)
file <- paste(a, "g2rm_high0exp_all_CTs.tsv", sep="/")
g2rm <- fread(file, stringsAsFactors = F, data.table = F) # 19309 transcripts
dat <- dat %>% filter(!ENSGID %in% g2rm$ENSGID)
rm(g2rm)

#- keep only protein-coding genes on autosomes (because h2 enrichment only considers autosomal SNPs)
genematrix <- fread(path_geneMx, 
                    stringsAsFactors = F, data.table = F) %>%
  filter(gene_type=="protein_coding") %>%
  filter(EMHC==FALSE) %>%
  filter(!hg19g0%in%c("chrM","chrX","chrY")) %>%
  filter(hg19g0!="" & is.na(g1)==F & is.na(g2)==F & gstr!="") %>%
  mutate(chr=hg19g0,
         start=as.integer(ifelse(g1-50000<0,0,g1-50000)), # hg19 coordinates, add 50kb window around each gene
         end=as.integer(g2+50000) ) %>%
  select(ensgid, chr, start, end)
g2rm <- dat %>% distinct(ENSGID) %>% filter(!ENSGID %in% genematrix$ensgid) 
dat <- dat %>% filter(!ENSGID %in% g2rm$ENSGID)
rm(g2rm)

#= 3. normalize to 1 TPM per cluster
dat <- dat %>% 
  group_by(cluster) %>% 
  mutate(cluster_TPM=Expr*1e6/sum(Expr)) %>% 
  ungroup()

#= 4. calculate specificity: normalized expression divided by sum of normalized expression across cell types
dat <- dat %>% 
  group_by(ENSGID) %>% 
  mutate(specificity=cluster_TPM/sum(cluster_TPM)) %>% 
  ungroup()

#= 5. map to hg19 coordinate 
dat <- dat %>% 
  inner_join(genematrix, by=c("ENSGID"="ensgid")) 
save(dat,file = "./5b_differential_gene_expression_analysis/output/speMx_Expr_15CTs.Rdata")

#= 6. output
valid.spe.long <- dat %>% filter(cluster_TPM>1) %>% # further remove genes with low expression (normalized expressionâ‰¤1)
  group_by(cluster) %>%
  mutate(n_genes=n(),
         n_genes_to_keep=round(n_genes*0.1)) %>%
  select(-n_genes)

#- function from Bryois (PMID=29785013)
ldsc_bedfile <- function(d,cluster){
  d_spe <- d %>% group_by_(cluster) %>% top_n(.,n_genes_to_keep,specificity) 
  d_spe %>% do(write_group(.,cluster))
}
write_group  = function(df,cluster) {
  df <- select(df,cluster,chr,start,end,ENSGID)
  dir.create(paste0("Pipeline_Specificity"), showWarnings = FALSE,recursive = TRUE)
  write_tsv(df[-1],paste0("Pipeline_Specificity/",make.names(unique(df[1])),".bed"),col_names = F)
  return(df)
}
# rm existing bed files
system("rm Pipeline_Specificity/*.bed")
# Note: output to current directory
valid.spe.long %>% ldsc_bedfile("cluster")

```
  
3. run h2 enrichment analysis.  
```{bash}
SCRIPTS_DIR=./7_heritability_enrichment/scripts
PIPELINE=Pipeline_Specificity
MAIN_DIR=./7_heritability_enrichment/data

cd ${MAIN_DIR}/${PIPELINE}

#- 1. calculate LD
sbatch -o ${MAIN_DIR}/${PIPELINE}/LOG_01a_LDSC_calculate_LD.log \
    -n 1 -t 01:00:00 --mem=16g \
    ${SCRIPTS_DIR}/01a_pLDSC_calculate_LD.sh ${PIPELINE}

#- 2. partition heritability (run when step 1 finishes)
sbatch -o ${MAIN_DIR}/${PIPELINE}/LOG_01b_LDSC_partition_h2.log \
    -n 1 -t 01:00:00 --mem=16g \
    ${SCRIPTS_DIR}/01b_pLDSC_partition_h2.sh ${PIPELINE}
```
  
4. clean results  
```{r}
library(data.table)
library(tidyverse)
library(stringr)

WD="./7_heritability_enrichment/data/Pipeline_Specificity/Results/" # path with .results file from partitioned LDSC analysis
outpath="./7_heritability_enrichment/output/"

#= LDSC results df
myfun.ldsc <- function(a){
files <- list.files(a,pattern=".results",full.names = TRUE)
d <- data_frame(filename=files) %>% mutate(file_contents = map(filename,read_tsv)) %>%
  mutate(Tissue=gsub(".bed_tissue_dir.results","",basename(filename))
         ) %>% unnest() %>%
  filter(Category=="L2_1") %>% mutate(P=1-pnorm(`Coefficient_z-score`)) %>% mutate(Trait=sub("_.*","",Tissue)) %>%
  mutate(Tissue=gsub("^_","",str_extract(Tissue, "_.*"))) %>%
  select(Trait,Tissue,P,Enrichment,Enrichment_std_error) %>%
  rename(VARIABLE=Tissue, BETA=Enrichment, SE=Enrichment_std_error)
write_tsv(d,path=paste(a,"/tissue_pvalues.txt", sep=""))
return(d)
}

ntest <- 16

df.ldsc.count2 <- myfun.ldsc(WD) %>% 
  mutate(if.sig=ifelse(P<=0.05/ntest, 1, 0))

df.ldsc.count2$VARIABLE <- gsub("\\.", "-", df.ldsc.count2$VARIABLE)
df.ldsc.count2$VARIABLE <- gsub("_", " ", df.ldsc.count2$VARIABLE)

d <- df.ldsc.count2 %>%
  arrange(-P) %>%
  mutate(order=row_number()) 

b <- d %>% distinct(VARIABLE,order, sig) %>% arrange(order) %>% select(sig)

library(readxl)

# read color code
a <- "./workflow/Cell_type_colors.xlsx" # file for color code
color.code <- read_excel(a)
dat.color <- color.code %>%
  distinct(`Cell type (15)`, `Cell type (15) color`) %>%
  rename(VARIABLE=`Cell type (15)`,
         ct_color=`Cell type (15) color`)

e <- d %>%
  left_join(dat.color, by="VARIABLE")

fwrite(e, 
       paste(outpath, "fig2F/df_fig2F_h2SpecificGenes.tsv", sep=""), 
       sep="\t", col.names = T)
```
  
-- end --  